python/kvikio/_lib/pynvcomp.pyx

```diff
@@ -0,0 +1,429 @@
+# Copyright (c) 2022 Carson Swope
+# Use, modification, and distribution is subject to the MIT License
+# https://github.com/carsonswope/py-nvcomp/blob/main/LICENSE)
+# SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: MIT
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+from enum import Enum
+
+import cupy as cp
+
+from libc.stdint cimport uintptr_t
+from libcpp cimport bool
+
+from kvikio._lib.pynvcomp cimport (
+    __CascadedCompressor,
+    __CascadedDecompressor,
+    __LZ4Compressor,
+    __LZ4Decompressor,
+    cudaStream_t,
+    nvcompBatchedSnappyCompressAsync,
+    nvcompBatchedSnappyCompressGetMaxOutputChunkSize,
+    nvcompBatchedSnappyCompressGetTempSize,
+    nvcompBatchedSnappyDecompressAsync,
+    nvcompBatchedSnappyDecompressGetTempSize,
+    nvcompBatchedSnappyOpts_t,
+    nvcompStatus_t,
+    nvcompType_t,
+)
+
+
+cpdef __get_array_interface_ptr(a):
+    return a.__array_interface__['data'][0]
+
+cpdef __get_cuda_array_interface_ptr(a):
+    return a.__cuda_array_interface__['data'][0]
+
+# could be either __array_interface__ or __cuda_array_interface__
+cpdef __get_ptr(a):
+    # this has to be slow though... is there a better way? try/catch?
+    d = a.__dir__()
+    if '__cuda_array_interface__' in d:
+        return __get_cuda_array_interface_ptr(a)
+    elif '__array_interface__' in d:
+        return __get_array_interface_ptr(a)
+    else:
+        raise AttributeError('Argument does not implement __cuda_array_interface__ or __array_interface__')  # NOQA: E501
```

Comment:
We could use [`Array`]( https://github.com/rapidsai/kvikio/blob/4a60dc0bf046023e8cbc7d9c39afc663bd407588/python/kvikio/_lib/arr.pyx#L66 ) to handle this. Would also handle computing `nbytes`. Though this could be follow on work after this PR